torch/csrc/autograd/engine.cpp

```diff
@@ -352,6 +340,26 @@ auto Engine::thread_main(GraphTask *graph_task) -> void {
       }
     }
   }
+
+  if (graph_task) {
+    graph_task->not_done_.notify_all();
+  }
+}
+
+void Engine::reentrant_thread_init() {
+  at::init_num_threads();
+  auto tp_shared= thread_pool_shared_;
+  while(true) {
+    std::unique_lock<std::mutex> lk(tp_shared->mutex_);
+    tp_shared->work_.wait(lk, [&tp_shared]{ return !tp_shared->graphtasks_queue_.empty();});
+    --thread_pool_shared_->num_workers_;
+    auto graph_task = tp_shared->graphtasks_queue_.front();
+    tp_shared->graphtasks_queue_.pop();
+    lk.unlock();
+    set_device(graph_task->owner_);
+    thread_main(graph_task);
+    ++thread_pool_shared_->num_workers_;
```

Comment:
Shouldn't this increment be under the lock? Furthermore, if you move this under the lock, you will now have successfully removed all unsynchronized accesses to `num_workers_`, and so it doesn't have to be atomic anymore.

Reply:
If you move this to the beginning of the loop, you'll have to delete that other increment in the code (but that seems cleaner to me)